sht/README.md

```diff
@@ -0,0 +1,98 @@
+# `sht` (SHell Test)
+
+Author tests in shell that integrate with `go test`.
+
+## Quickstart
+
+```shell
+apk add sht
+```
+
+### Create a test file (`test.sh`)
+
+```bash
+#!/usr/bin/env sht
+# Use the `sht` interp to "wrap" the shell script
+
+# Test functions are defined as local functions with the `shtest_*` prefix
+shtest_hello() {
+  echo "hello world"
+}
+
+echo "this is just a regular shell script"
+
+shtest_fail() {
+  echo "any non-zero exit code will fail the test immediately (like set -e)"
+
+  cat foo
+
+  echo "we'll never get here"
+}
+```
+
+### Run the test
+
+```bash
+$ chmod +x test.sh
+
+$ ./test.sh
+=== RUN   TestShell
+=== RUN   TestShell/simple.sh
+    main_test.go:177: rendered wrapped script to: /var/folders/5v/6gvb9x954sbd9tmqq14cwgrh0000gn/T/sht-138616544/wsht.sh
+this is just a regular shell script
+=== RUN   TestShell/simple.sh/shtest_hello
+hello world
+    testfn.go:85: [shtest_hello] finished successfully
+=== RUN   TestShell/simple.sh/shtest_fail
+any non-zero exit code will fail the test immediately (like set -e)
+cat: foo: No such file or directory
+    testfn.go:90: [shtest_fail] finished with error code 1
+    main_test.go:189: failed to run test: test finished with error code 1
+--- FAIL: TestShell (0.28s)
+    --- FAIL: TestShell/simple.sh (0.28s)
+        --- PASS: TestShell/simple.sh/shtest_hello (0.04s)
+        --- FAIL: TestShell/simple.sh/shtest_fail (0.03s)
+FAIL
+```
+
+## How it works
+
+`sht` consists of three components:
+
+- **shtr**: The test runner
+- **shtc**: The test runner client
+- **sht**: A simple wrapper to support OS's without support for `env -S`
+
+On execution of a shell script using the `sht` interpreter, `shtr` "wraps" the
+script with the ["shell framework"](./framework.sh.tpl), and then executes it
+with `os/exec` using the framework's shbang (`#!/bin/sh`).
+
+The "framework" is mostly shell functions with a very thin amount of go templating
+to ensure all the appropriate pipes are configured.
+
+From there, the final rendered script is written to disk and executed by `shtr`.
+
+Communication between the shell script for stdout/stderr, and progress updates
+happen over named pipes, which `shtr` and `shtc` orchestrate.
+
+There are no modifications made to the original script, which means all of its
+contents are still executed. The only "trickery" done by `sht` is parsing (via
+`shfmt`) and explicitly calling all `shtest_*` functions with the appropriate
+`set -e` arguments.
+
+Each `shtest_*` function is run serially, in the order it is defined. Tests
+`PASS` when the `shtest_*` function exits with a 0 exit code, and tests `FAIL`
```

Comment:
exits ? or returns?
what does happen if the function itself `exit 1`, does it somehow escape the infrastructure?

Reply:
returns thanks.

each `shtest_*` is in its own subshell, so exits/returns just exit the subshell not the infra

Reply:
added a note about this